docs(compilation-hooks): improve descriptions and formatting (#2852)
webpack/webpack.js.org: src/content/api/compilation-hooks.md

Fired at the beginning of dependency optimization.

- Callback Parameters: `modules`

### `optimizeDependenciesAdvanced`

`SyncBailHook`

W> This hook will be removed in v5.0.0

- Callback Parameters: `modules`

### `afterOptimizeDependencies`

`SyncHook`

Fired after the dependency optimization.

- Callback Parameters: `modules`

### `optimize`

`SyncHook`

> Triggered at the beginning of the optimization phase.  

Reviewer: ```suggestion
Triggered at the beginning of the optimization phase.
```

### `optimizeModulesBasic`

`SyncBailHook`

W> This hook will be removed in v5.0.0

- Callback Parameters: `modules`

### `optimizeModules`

`SyncBailHook`

> Called at the beginning of the modules optimization phase. A plugin can tap into  this hook to perform modules optimizations.

Reviewer: ```suggestion
Called at the beginning of the modules optimization phase. A plugin can tap into this hook to perform modules optimizations.
```

- Callback Parameters: `modules`

### `optimizeModulesAdvanced`

`SyncBailHook`

W> This hook will be removed in v5.0.0

- Callback Parameters: `modules`

### `afterOptimizeModules`

`SyncHook`

> Fired after modules optimization has completed.

Reviewer: ```suggestion
Called after modules optimization has completed.
```

Fired after chunk optimization has completed.

- Callback Parameters: `chunks`

### `optimizeTree`

`AsyncSeriesHook`

Called before optimizing the dependency tree. A plugin can tap into this hook to perform a dependency tree optimization.

- Callback Parameters: `chunks` `modules`

### `afterOptimizeTree`

`SyncHook`

> Fired after the dependency three optimization has completed with success.

Reviewer: ```suggestion
Called after the dependency tree optimization has completed with success.
```

- Callback Parameters: `chunks` `modules`

### `optimizeChunkModulesBasic`

`SyncBailHook`

W> This hook will be removed in v5.0.0

- Callback Parameters: `chunks` `modules`

### `optimizeChunkModules`

`SyncBailHook`

> Called after the three optimization, at the beginning of the chunkmodules optimization. A plugin can tap into this hook to perform a chunkmodules optimization.

Reviewer: Can you link `chunkmodules` please

Author: Sure, looking in the documentation I can't find `chunkmodules`, do you mean the [code splitting](https://webpack.js.org/guides/code-splitting/) section, the stats section or something else? 

Thanks a lot for the corrections!

Reviewer: I mean, it doesn't read right as `chunkmodules`, looks like it would run after the tree optimization indeed, and not three. Lets try to change it to:
```suggestion
Called after the tree optimization, at the beginning of the chunk modules optimization. A plugin can tap into this hook to perform optimizations of chunk modules.
```

- Callback Parameters: `chunks` `modules`

### `optimizeChunkModulesAdvanced`

`SyncBailHook`

W> This hook will be removed in v5.0.0

- Callback Parameters: `chunks` `modules`

### `afterOptimizeChunkModules`

`SyncHook`

> Fired after the chunkmodules optimization has completed with success.

Reviewer: ```suggestion
Called after the chunkmodules optimization has successfully completed.
```

- Callback Parameters: `chunks` `modules`

### `shouldRecord`

`SyncBailHook`

> Called to determine wheter or not storing records. Returning anything `!== false` will prevent every other "record" hook from being executed (`record`, `recordModules`, `recordChunks` and `recordHash` )

Reviewer: ```suggestion
Called to determine whether or not to store records. Returning anything `!== false` will prevent every other "record" hook from being executed (`record`, `recordModules`, `recordChunks` and `recordHash` )
```

Restore module information from records.

- Callback Parameters: `modules` `records`

### `optimizeModuleOrder`

`SyncHook`

W> This hook will be removed in v5.0.0

Sort the modules in from most to least important.

- Callback Parameters: `modules`

### `advancedOptimizeModuleOrder`

`SyncHook`

W> This hook will be removed in v5.0.0

- Callback Parameters: `modules`

### `beforeModuleIds`

`SyncHook`

Executed before assigning an `id` to each module.

- Callback Parameters: `modules`

### `moduleIds`

`SyncHook`

Called to assign an `id` to each module.

- Callback Parameters: `modules`

### `optimizeModuleIds`

`SyncHook`

> Fired at the beginning of the modules `id` optimization.

Reviewer: ```suggestion
Called at the beginning of the modules `id` optimization.
```

Restore chunk information from records.

- Callback Parameters: `chunks` `records`

### `optimizeChunkOrder`

`SyncHook`

W> This hook will be removed in v5.0.0

Sort the chunks in from most to least important.

- Callback Parameters: `chunks`
  

### `beforeChunkIds`

`SyncHook`

> Executed before assigning an `id` to each chunks.

Reviewer: ```suggestion
Executed before assigning an `id` to each chunk.
```

- Callback Parameters: `chunks`

### `chunkIds`

`SyncHook`

T> This hook will be available in v5.0.0

> Called to assign an `id` to each chunks.

Reviewer: ```suggestion
Called to assign an `id` to each chunk.
```

- Callback Parameters: `modules`

### `beforeOptimizeChunkIds`

`SyncHook`

T> This hook will be available in v5.0.0

Fired before chunks `id` optimization.

- Callback Parameters: `chunks`

### `optimizeChunkIds`

`SyncHook`

> Fired at the beginning of the chunks `id` optimization phase.

Reviewer: ```suggestion
Called at the beginning of the chunks `id` optimization phase.
```

Triggered after chunk `id` optimization has finished.

- Callback Parameters: `chunks`

### `recordModules`

`SyncHook`

> Store module info to the records. This is triggered only if 

Reviewer: Only if?

Store chunk info to the records.

- Callback Parameters: `chunks` `records`

### `optimizeCodeGeneration`

T> This hook will be available in v5.0.0

A plugin can tap into this hook to optimize the generated code.

- Callback Parameters: `modules`

> ### `beforeModuleHash`

Reviewer: This and few following hooks are missing callback parameters

Author: That's because they don't pass parameters to the callbacks. We can write `- callback parameters: none` but I don't like it much

Reviewer: Great, just confirming!

Store information about the `compilation` to the `records`.

- Callback Parameters: `compilation` `records`

### `beforeModuleAssets`

`SyncHook`

Executed before module assets creation.

### `additionalChunkAssets`

`SyncHook`

Create additional assets for the chunks.

- Callback Parameters: `chunks`

### `shouldGenerateChunkAssets`

`SyncBailHook`

> Called to determine wheter or not generate chunk assets. Returning anything `!== false` will allow chunk assets generation.

Reviewer: ```suggestion
Called to determine wheter or not generate chunks assets. Returning anything `!== false` will allow chunk assets generation.
```

### `beforeChunkAssets`

`SyncHook`

> Executed before creating the chunk assets.

Reviewer: ```suggestion
Executed before creating the chunks assets.
```
